File: utils/integrity.cpp
// utils/integrity.cpp — SHA-256 file integrity checking
// Equivalent of the Python utils/integrity.py

#include "integrity.h"

#include <QFile>
#include <QCryptographicHash>

namespace Integrity {

QString generateFileHash(const QString &filePath) {
    QFile file(filePath);
    if (!file.open(QIODevice::ReadOnly)) return QString();

    QCryptographicHash hash(QCryptographicHash::Sha256);
    constexpr qint64 chunkSize = 8192;
    QByteArray chunk;
    while (!(chunk = file.read(chunkSize)).isEmpty()) {
        hash.addData(chunk);
    }
    return QString::fromLatin1(hash.result().toHex());
}

bool checkIntegrity(const QStringList &filePaths, const QStringList &expectedHashes) {
    if (filePaths.size() != expectedHashes.size()) return false;
    for (int i = 0; i < filePaths.size(); ++i) {
        QString actual = generateFileHash(filePaths.at(i));
        if (actual.isEmpty() || actual != expectedHashes.at(i))
            return false;
    }
    return true;
}

}  // namespace Integrity

File: utils/integrity.py
import os
import hashlib

def generate_file_hash(filepath):
    """Generate SHA-256 hash of a file."""
    if not os.path.exists(filepath):
        return None
    sha256 = hashlib.sha256()
    with open(filepath, 'rb') as f:
        while chunk := f.read(8192):
            sha256.update(chunk)
    return sha256.hexdigest()

def check_integrity(file_paths, expected_hashes):
    """
    Check if a list of files matches their expected hashes.
    Expected to be expanded to read a lockfile (e.g. integrity.lock)
    """
    for file, expected in zip(file_paths, expected_hashes):
        actual = generate_file_hash(file)
        if actual != expected:
            return False
    return True
